fieldpath/set_test.go

```diff
@@ -755,3 +756,255 @@ func TestSetNodeMapIterate(t *testing.T) {
 		}
 	}
 }
+
+func TestFilterByPattern(t *testing.T) {
+	testCases := []struct {
+		name   string
+		input  *Set
+		expect *Set
+		filter Filter
+	}{
+		{
+			name: "container resize fields: exact match",
+			input: NewSet(
+				MakePathOrDie("spec"),
+				MakePathOrDie("spec", "containers"),
+				MakePathOrDie("spec", "containers", 0, "resources"),
+				MakePathOrDie("spec", "containers", 0, "resources", "limits"),
+				MakePathOrDie("spec", "containers", 0, "resources", "limits", "cpu"),
+				MakePathOrDie("spec", "containers", 0, "resources", "requests"),
+				MakePathOrDie("spec", "containers", 0, "resources", "requests", "cpu"),
+				MakePathOrDie("spec", "containers", 0, "resources", "claims"),
+				MakePathOrDie("spec", "containers", 0, "resources", "claims", 0, "name"),
+				MakePathOrDie("spec", "containers", 0, "resources", "claims", 0, "request"),
+				MakePathOrDie("spec", "containers", 0, "resources", "claims", 1, "name"),
+				MakePathOrDie("spec", "containers", 0, "resources", "claims", 1, "request"),
+				MakePathOrDie("spec", "containers", 1, "resources"),
+				MakePathOrDie("spec", "containers", 1, "resources", "limits"),
+				MakePathOrDie("spec", "containers", 1, "resources", "limits", "cpu"),
+			),
+			filter: NewIncludeMatcherFilter(MakePrefixMatcherOrDie("spec", "containers", MatchAnyPathElement(), "resources")),
+			expect: NewSet(
+				MakePathOrDie("spec"),
+				MakePathOrDie("spec", "containers"),
+				MakePathOrDie("spec", "containers", 0, "resources"),
+				MakePathOrDie("spec", "containers", 0, "resources", "limits"),
+				MakePathOrDie("spec", "containers", 0, "resources", "limits", "cpu"),
+				MakePathOrDie("spec", "containers", 0, "resources", "requests"),
+				MakePathOrDie("spec", "containers", 0, "resources", "requests", "cpu"),
+				MakePathOrDie("spec", "containers", 0, "resources", "claims"),
+				MakePathOrDie("spec", "containers", 0, "resources", "claims", 0, "name"),
+				MakePathOrDie("spec", "containers", 0, "resources", "claims", 0, "request"),
+				MakePathOrDie("spec", "containers", 0, "resources", "claims", 1, "name"),
+				MakePathOrDie("spec", "containers", 0, "resources", "claims", 1, "request"),
+				MakePathOrDie("spec", "containers", 1, "resources"),
+				MakePathOrDie("spec", "containers", 1, "resources", "limits"),
+				MakePathOrDie("spec", "containers", 1, "resources", "limits", "cpu"),
+			),
+		},
+		{
+			name: "container resize fields: filter status and metadata",
+			input: NewSet(
+				MakePathOrDie("spec"),
+				MakePathOrDie("status"),
+				MakePathOrDie("metadata"),
+			),
+			filter: NewIncludeMatcherFilter(MakePrefixMatcherOrDie("spec", "containers", MatchAnyPathElement(), "resources")),
+			expect: NewSet(
+				MakePathOrDie("spec"),
```

Comment:
Semantic question - is this the right result?  Spec isn't a "leaf" and it didn't actually match the filter (it matched part of the filter).  I don't know whis codebase well enough, but that was a little surprising to me.  Yay for tests.

Reply:
The background here is that these filters are operating against SSA managed fields.  BUT, the filters are not aware of the SSA field granularity (the atomic vs. granular thing).  Worse, the SSA field granularity could change and it would probably be better if this is durable to such changes.

If we include parent fields in matches, then we're saying: "I don't care what the SSA granularity is, I logically want these fields kept, and if the managed fields cause the field set to terminate at some parent because it's atomic, that's fine."

Reply:
Yes, that's correct.